### ISSUE TITLE
Spinlock implementation is susceptible to priority inversion deadlocks

### ISSUE THREAD

**Issue by valkum** [xacrimon/dashmap#142]
We had a discussion where dashmap was mentioned and later [Spinlocks Considered Harmful by Aleksey Kladov](https://matklad.github.io/2020/01/02/spinlocks-considered-harmful.html) was shared. We came to the conclusion that dashmap is susceptible to the priority inversion problem under heavy contention.

In lock.rs multiple places use spinning in combination with `std::sync::atomic::spin_loop_hint` as cpu_relax.
E.g.
https://github.com/xacrimon/dashmap/blob/1a31f4f5989fa42f5faa80ae0537e214d55cb0e4/src/lock.rs#L137-L144
This function seems to be superseded by `std::hint::spin_loop`.
The documentation of `spin_loop` says clearly that you should not spin endlessly.

> To avoid problems like priority inversion, it is strongly recommended that the spin loop is terminated after a finite amount of iterations and an appropriate blocking syscall is made.

The linked article (https://matklad.github.io/2020/01/02/spinlocks-considered-harmful.html), explains the problem in detail.

**Comment by rsdy** [comment]
I have ran into this issue with a lot of tokio tasks accessing the store in parallel. It's probably not very difficult to reproduce in an isolated test-case.

**Comment by xacrimon** [comment]
Yep, I am aware. At the moment spinlocks are used because parking_lot and std didn't have the needed features when the code was written. A parking_lot feature is in the works.

**Comment by ljw1004** [comment]
For reference:
PR https://github.com/xacrimon/dashmap/pull/145 "feat: add parking_lot as a feature"
issue https://github.com/xacrimon/dashmap/issues/137 "Feature request: Use parking_lot for RwLock"

**Comment by xacrimon** [comment]
Fixed as of v5
